### PR TITLE
Phase 6: Fourier Feature Position Encoding — Spectral Bias Correction

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Neural networks have a well-known **spectral bias**: they preferentially fit low-frequency components of a function before high-frequency ones (Rahaman et al., ICML 2019). This is a problem for sharp pressure peaks at airfoil leading edges and suction peaks, which require high-frequency spatial representations.

**Random Fourier Features (Tancik et al., NeurIPS 2020)** maps input positional coordinates through random sinusoidal projections before passing them to the network:

```
γ(x) = [sin(2π B x), cos(2π B x)]
```

where B is sampled from a Gaussian with tunable scale σ. This is the same key insight behind positional encoding in Transformers (Vaswani et al.) and NeRF (Mildenhall et al., 2020) — it lifts the coordinate representation into a higher-frequency basis, allowing the network to fit sharp spatial features without spectral bias.

**Why this matters here:** The model struggles most with:
- p_tan = 29.1 — aft-foil suction peaks in tandem configurations (sharp, localized)
- p_in = 12.1 — single-foil leading-edge suction peaks

Both are characterized by very sharp pressure gradients in small spatial regions. The current model's L1 surface loss on these sharp features is limited by the network's tendency to under-predict peak magnitudes. Richer frequency encoding of spatial positions should help the model "see" these sharp features more clearly.

**Key distinction from augmentation:** This is not a data augmentation — it's a change to how spatial coordinates are represented to the model. The Fourier features replace or augment the raw (x, y) coordinate channels in the node features with a higher-dimensional representation that spans multiple frequency bands.

**Paper evidence:** Tancik et al. (NeurIPS 2020) demonstrated that Fourier feature networks achieve 10-40x lower error on high-frequency regression tasks vs standard MLPs. The benefit was largest for tasks with sharp spatial features (exactly our case). Mildenhall et al. (NeRF) used it for 3D scene reconstruction — we're doing 2D pressure field reconstruction.

## Instructions

The change adds Fourier feature encoding to the positional coordinate channels of the input features.

**Step 1: Add `--fourier_features` and `--fourier_scale` flags** to argparse:
```python
parser.add_argument('--fourier_features', type=int, default=0,
                    help='Number of Fourier feature frequency bands (0=disabled, try 16 or 32)')
parser.add_argument('--fourier_scale', type=float, default=10.0,
                    help='Gaussian scale for Fourier feature sampling (controls frequency range)')
```
Add to Config dataclass:
```python
fourier_features: int = 0      # 0 = disabled
fourier_scale: float = 10.0   # frequency scale
```

**Step 2: Create the Fourier feature projection matrix** (fixed random, registered as a buffer):

In the Transolver model `__init__`:
```python
if cfg.fourier_features > 0:
    # Sample fixed random projection matrix: [2, fourier_features]
    # The 2 input dims are the (x, y) spatial coordinates
    B_matrix = torch.randn(2, cfg.fourier_features) * cfg.fourier_scale
    self.register_buffer('fourier_B', B_matrix)  # fixed (no gradient)
    # Project to match n_hidden input dim — add a small linear to integrate with existing encoding
    self.fourier_proj = nn.Linear(2 * cfg.fourier_features, cfg.n_hidden, bias=True)
    nn.init.zeros_(self.fourier_proj.weight)
    nn.init.zeros_(self.fourier_proj.bias)
```

Zero-init ensures the model starts from baseline behavior and learns to use Fourier features gradually.

**Step 3: Apply in the model's forward pass**, after the position embedding but before the first TransolverBlock:

Find where `x` (node features) is first embedded in the model, and locate the position / coordinate channels. In Transolver, spatial positions are typically passed as `pos` or included in the first few channels of `x`.

```python
if self.cfg.fourier_features > 0 and self.fourier_B is not None:
    # pos: [B, N, 2] — (x, y) coordinates of mesh nodes
    # Project: [B, N, 2] @ [2, F] = [B, N, F]
    proj = (2 * torch.pi * pos) @ self.fourier_B  # [B, N, fourier_features]
    fourier_enc = torch.cat([torch.sin(proj), torch.cos(proj)], dim=-1)  # [B, N, 2F]
    # Add to existing node embedding as a residual
    fx = fx + self.fourier_proj(fourier_enc)  # [B, N, n_hidden]
```

The zero-init of `fourier_proj` means the model starts as the current baseline.

**Step 4: Find where `pos` is available.** Check how the model receives coordinates. If `pos` is part of `x` (the first 2 channels), extract it: `pos = x[:, :, :2]` or wherever the (x, y) node coordinates are stored. Look at the existing code for where `pos` is used in the embedding step.

**Experiment runs** (try two scales with 2 seeds each):

```bash
# 16 features, scale=10, seed 42
python train.py --agent askeladd \
  --wandb_name "askeladd/ff-f16-sc10-s42" \
  --wandb_group phase6/fourier-features \
  --fourier_features 16 --fourier_scale 10.0 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# 16 features, scale=10, seed 73
python train.py --agent askeladd \
  --wandb_name "askeladd/ff-f16-sc10-s73" \
  --wandb_group phase6/fourier-features \
  --fourier_features 16 --fourier_scale 10.0 --seed 73 \
  [same flags as above]

# 32 features, scale=10, seed 42
python train.py --agent askeladd \
  --wandb_name "askeladd/ff-f32-sc10-s42" \
  --wandb_group phase6/fourier-features \
  --fourier_features 32 --fourier_scale 10.0 --seed 42 \
  [same flags as above]

# 32 features, scale=10, seed 73
python train.py --agent askeladd \
  --wandb_name "askeladd/ff-f32-sc10-s73" \
  --wandb_group phase6/fourier-features \
  --fourier_features 32 --fourier_scale 10.0 --seed 73 \
  [same flags as above]
```

If you have VRAM headroom, run all 4 in parallel. The Fourier feature projection adds negligible parameters (<1% overhead).

W&B group: `phase6/fourier-features`

## Baseline

Current best (16-seed ensemble, PR #2093):

| Metric | Baseline |
|--------|----------|
| p_in   | **12.1** |
| p_oodc | **6.6**  |
| p_tan  | **29.1** |
| p_re   | **5.8**  |

Single-model references:
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Individual Run Metrics

| Config | Seed | val/loss | p_in | p_oodc | p_tan | p_re | W&B Run |
|--------|------|----------|------|--------|-------|------|---------|
| FF=16, σ=10 | 42 | 0.3939 | 13.6 | 8.1 | 29.8 | 6.9 | i36jee3d |
| FF=16, σ=10 | 73 | 0.4002 | 14.1 | 8.2 | 31.1 | 6.6 | 28jblql6 |
| FF=32, σ=10 | 42 | 0.3921 | 12.9 | 8.1 | 30.0 | 6.8 | bkktfeu6 |
| FF=32, σ=10 | 73 | 0.3942 | 13.2 | 8.5 | 29.3 | 6.9 | snipvwml |

### Mean by Config

| Config | val/loss | p_in | p_oodc | p_tan | p_re |
|--------|----------|------|--------|-------|------|
| **Baseline** (8-seed mean) | **0.3829** | **13.03** | **7.83** | **30.29** | **6.45** |
| FF=16, σ=10 (2-seed) | 0.3971 | 13.85 | 8.15 | 30.45 | 6.75 |
| FF=32, σ=10 (2-seed) | 0.3932 | 13.05 | 8.30 | **29.65** | 6.85 |

### Key Metric Comparison (vs 8-seed baseline)

| Metric | Baseline | FF=16 | Δ | FF=32 | Δ |
|--------|----------|-------|---|-------|---|
| p_in | 13.03 | 13.85 | +6.3% | 13.05 | +0.2% |
| p_oodc | 7.83 | 8.15 | +4.1% | 8.30 | +6.0% |
| p_tan | 30.29 | 30.45 | +0.5% | **29.65** | **-2.1%** |
| p_re | 6.45 | 6.75 | +4.7% | 6.85 | +6.2% |

### Same-Seed Comparison (seed 42)

| Metric | Baseline s42 | FF=16 s42 | FF=32 s42 |
|--------|-------------|-----------|-----------|
| p_in | ~12.7 | 13.6 (+7.1%) | 12.9 (+1.6%) |
| p_oodc | ~8.1 | 8.1 (0%) | 8.1 (0%) |
| p_tan | ~30.7¹ | 29.8 (-2.9%) | 30.0 (-2.3%) |

¹ Recent s42 baseline run

### Implementation Details

Added random Gaussian Fourier feature encoding to the Transolver model:
- Fixed random projection matrix `B ~ N(0, σ²)` of shape `[2, fourier_features]`
- Node (x,y) coordinates → `sin(2π·pos·B), cos(2π·pos·B)` → `[B, N, 2F]`
- Zero-initialized linear projection to `n_hidden` dimensions
- Added as a residual to `fx` after `preprocess`, before TransolverBlocks

This operates in the hidden space (not input features), complementing the existing 8-frequency Fourier PE already appended to the input.

### Commands Used

```bash
CUDA_VISIBLE_DEVICES={0,1,2,3} python train.py --agent askeladd \
  --wandb_group phase6/fourier-features \
  --fourier_features {16,32} --fourier_scale 10.0 --seed {42,73} \
  --wandb_name "askeladd/ff-f{16,32}-sc10-s{42,73}" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

W&B group: `phase6/fourier-features`

### What Happened

**Mixed results — p_tan improves but p_oodc and p_re degrade.**

The FF=32 configuration shows a promising **-2.1% improvement on p_tan** (29.65 vs 30.29), with p_in essentially matching baseline (13.05 vs 13.03). However, p_oodc degrades by 6.0% and p_re by 6.2%.

FF=16 performs worse overall — the smaller number of frequency bands doesn't provide enough spectral coverage to help, while still introducing enough noise to hurt p_in.

The p_tan improvement is consistent across both seeds (29.8 and 29.3 for FF=32's s42 and s73 respectively), suggesting the Fourier features genuinely help the model resolve sharp pressure features in tandem configurations. However, the OOD condition and Re splits lose accuracy, possibly because the random Fourier frequencies introduce position-dependent biases that don't generalize well across different flow conditions.

### Analysis

The hypothesis about spectral bias correction is partially validated:
- **p_tan improves**: Tandem foils have the sharpest pressure interactions, and richer position encoding helps
- **p_in is neutral**: Single-foil suction peaks are already well-captured by existing encoding
- **p_oodc/p_re degrade**: The fixed random frequencies may overfit to position patterns that are geometry-specific rather than physics-general

The model already has 8-frequency Fourier PE in the input (4 fixed + 4 learnable). Adding more in the hidden space provides modest benefit for tandem transfer but hurts generalization on other OOD splits. The existing input-level PE may already be sufficient.

### Suggested Follow-ups

- **FF=32 warrants multi-seed validation** if the advisor sees the p_tan improvement as worth pursuing despite p_oodc/p_re regression
- **Learnable Fourier frequencies** (instead of fixed random) in the hidden space might help — the model could learn to focus on frequencies that generalize
- **Lower scale (σ=1-5)** could produce frequencies that are more relevant to the physical length scales in the data
- The regression on p_oodc/p_re may be fixable with a smaller residual scale (e.g., `0.1 * fourier_proj(enc)` instead of `1.0x`)

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2106 because Results show mixed outcomes. FF=32 (σ=10, 2-seed mean): p_tan -1.3% (29.65 vs 30.05 ✓) and p_in -1.1% (13.05 vs 13.19 ✓), but p_oodc +4.8% (8.30 vs 7.92 ✗) and p_re +6.2% (6.85 vs 6.45 ✗). The p_tan signal is promising but the generalization regressions on OOD and Re splits disqualify this as a merge. As the student correctly identified, the model already has 8-frequency Fourier PE in the input — the hidden-space addition duplicates spectral coverage without sufficient benefit. FF=16 is strictly worse across all metrics. Closing as a dead end. The student's analysis was thorough and the implementation clean. Key learning: hidden-space Fourier encoding does not add value when input-level PE is already rich..
